exercises/simple-linked-list/src/lib.rs

```diff
@@ -0,0 +1,36 @@
+pub struct SimpleLinkedList<T> {
+    // Delete this field
+    // _dummy is needed to avoid unused parameter error during compilation
+    _dummy: T
+    // head
+    // size
+}
+
+// Implementation of basic functions
+// Use SimpleLinkedList as Stack (LIFO - Last In, First Out)
+impl<T> SimpleLinkedList<T> {
+    pub fn new() -> Self {
+        unimplemented!()
+    }
+
+    pub fn size(&self) -> usize {
+        unimplemented!()
+    }
+
+    pub fn push(&mut self, element: T) {
+        unimplemented!()
+    }
+
+    pub fn pop(&mut self) -> Option<T> {
+        unimplemented!()
+    }
+
+    pub fn peek(&self) -> Option<&T> {
+        unimplemented!()
+    }
+
+    // This is where you might consider to bind the generic type parameter with a trait
+    pub fn from_slice(slice: &[T]) -> Self {
```

Comment:
```
75 | impl<T> From<[T]> for SimpleLinkedList<T>
   |         ^^^^^^^^^ `[T]` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `[T]`
```
This is way I didn't implemented it

Reply:
Right, sorry; the impl has to be [a little more complex](https://play.rust-lang.org/?gist=03825bfd7841ffc4eb054e1913b6f5d0&version=stable):

```rust
impl<'a, T> From<&'a [T]> for SimpleLinkedList<T> where T: Clone  {
    fn from(_: &[T]) -> SimpleLinkedList<T> {
        unimplemented!()
    }
}
```

The general point, which is that for pedagogical rigor we should use appropriate traits instead of dedicated methods, stands.

Reply:
I had this idea, too. I've already implemented it, but when testing it the following error occurs:
```
58 |     let mut list = SimpleLinkedList::from(&slice);
   |                    ^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<&[&str; 4]>` is not implemented for `simple_linked_list::SimpleLinkedList<_>`
   |
   = help: the following implementations were found:
             <simple_linked_list::SimpleLinkedList<T> as std::convert::From<&'a [T]>>
   = note: required by `std::convert::From::from`
```
So I think `fn from_slice` is the better solution

Reply:
Arrays and slices are not quite the same thing. 

There are a few ways around this. We could improve the trait's bounds to cover what we're after, [which works](https://play.rust-lang.org/?gist=0b5560c45499c0339093a60e17f58a75&version=stable):

```rust
impl<'a, S, T> From<&'a S> for SimpleLinkedList<T>
where S: AsRef<[T]>,
      T: Clone,      
{
    fn from(_: &S) -> SimpleLinkedList<T> {
        unimplemented!()
    }
}
```

If we were writing a production library, that's probably the path we'd take. However, that might be a bit intimidating for the student, even if we gave them the stub. It's friendlier to them to use another approach. Luckily, that's a pretty simple matter of [changing the calling pattern](https://play.rust-lang.org/?gist=42757091d364b1d930ce1f91f5a18ab2&version=stable): `let _: SimpleLinkedList<_> = [1, 2, 3].as_ref().into();` works even with the previous, simpler impl.